social_core/backends/apple.py

```diff
@@ -0,0 +1,131 @@
+"""
+
+import json
+import time
+
+import jwt
+from jwt.algorithms import RSAAlgorithm
+from jwt.exceptions import PyJWTError
+from social_core.backends.oauth import BaseOAuth2
+from social_core.exceptions import AuthCanceled
+
+
+class AppleIdAuth(BaseOAuth2):
+    name = "apple-id"
+
+    JWK_URL = "https://appleid.apple.com/auth/keys"
+    AUTHORIZATION_URL = "https://appleid.apple.com/auth/authorize"
+    ACCESS_TOKEN_URL = "https://appleid.apple.com/auth/token"
+    ACCESS_TOKEN_METHOD = "POST"
+    RESPONSE_MODE = None
+
+    ID_KEY = "sub"
+    TOKEN_KEY = "id_token"
+    STATE_PARAMETER = True
+    REDIRECT_STATE = False
+
+    TOKEN_AUDIENCE = "https://appleid.apple.com"
+    TOKEN_TTL_SEC = 6 * 30 * 24 * 60 * 60
+
+    def auth_params(self, *args, **kwargs):
+        """Apple requires to set `response_mode` to `form_post` if `scope` parameter is passed."""
+        params = super(AppleIdAuth, self).auth_params(*args, **kwargs)
+        if self.RESPONSE_MODE:
+            params["response_mode"] = self.RESPONSE_MODE
+        elif self.get_scope():
+            params["response_mode"] = "form_post"
+        return params
+
+    def get_private_key(self):
+        """
+        Return contents of the private key file.
+        Override this method to provide secret key from another source if needed.
+        """
+        return self.setting("SECRET")
+
+    def generate_client_secret(self):
+        now = int(time.time())
+        client_id = self.setting("CLIENT")
+        team_id = self.setting("TEAM")
+        key_id = self.setting("KEY")
+        private_key = self.get_private_key()
+
+        headers = {"kid": key_id}
+        payload = {
+            "iss": team_id,
+            "iat": now,
+            "exp": now + self.TOKEN_TTL_SEC,
+            "aud": self.TOKEN_AUDIENCE,
+            "sub": client_id,
+        }
+
+        return jwt.encode(payload, key=private_key, algorithm="ES256", headers=headers)
+
+    def get_key_and_secret(self):
+        client_id = self.setting("CLIENT")
+        client_secret = self.generate_client_secret()
+        return client_id, client_secret
+
+    def get_apple_jwk(self):
+        keys = self.get_json(url=self.JWK_URL).get("keys")
+        if not isinstance(keys, list) or not keys:
+            raise AuthCanceled("Invalid jwk response")
+
+        return json.dumps(keys.pop())
+
+    def decode_id_token(self, id_token):
+        """Decode and validate JWT token from apple and return payload including user data."""
+        if not id_token:
+            raise AuthCanceled("Missing id_token parameter")
+
+        public_key = RSAAlgorithm.from_jwk(self.get_apple_jwk())
+        try:
+            decoded = jwt.decode(id_token, key=public_key, audience=self.setting("CLIENT"), algorithm="RS256",)
+        except PyJWTError:
+            raise AuthCanceled("Token validation failed")
+
+        return decoded
+
+    def get_user_details(self, response):
+        name = response.get("name") or {}
```

Comment:
I'm not sure if I'm misusing it, but when I implemented this verbatim in my app, `response` appeared to be the initial oauth response from apple and did not have the user fields.

I ended up changing this so that response was instead defined as

```
response = json.loads(self.data.get('user'))
```

From there, the rest of the name parsing logic matched up.

Reply:
Ah - I had it backwards. `response` is the decoded id_token, which [does not include the name anymore](https://forums.developer.apple.com/thread/124169).

Given email/name all comes back in the initial post (`self.data`), it almost feels like there is little need to do go beyond the `auth_complete` step since the post data with the `code` already includes the user info fields.